rust/agama-network/src/dbus/tree.rs

```diff
@@ -0,0 +1,229 @@
+        )
+        .await?;
+
+        Ok(())
+    }
+
+    /// Clears all the connections from the tree.
+    async fn remove_connections(&self) -> Result<(), ServiceError> {
+        let mut objects = self.objects.lock();
+        for path in objects.connections.values() {
+            self.remove_connection_on(path.as_str()).await?;
+        }
+        objects.connections.clear();
+        Ok(())
+    }
+
+    /// Clears all the devices from the tree.
+    async fn remove_devices(&mut self) -> Result<(), ServiceError> {
+        let object_server = self.connection.object_server();
+        let mut objects = self.objects.lock();
+        for path in objects.devices.values() {
+            object_server
+                .remove::<interfaces::Device, _>(path.as_str())
+                .await?;
+        }
+        objects.devices.clear();
+        Ok(())
+    }
+
+    /// Removes a connection object on the given path
+    ///
+    /// * `path`: connection D-Bus path.
+    async fn remove_connection_on(&self, path: &str) -> Result<(), ServiceError> {
+        let object_server = self.connection.object_server();
+        _ = object_server.remove::<interfaces::Wireless, _>(path).await;
+        object_server.remove::<interfaces::Ipv4, _>(path).await?;
+        object_server
+            .remove::<interfaces::Connection, _>(path)
+            .await?;
+        Ok(())
+    }
+
+    async fn add_interface<T>(&self, path: &str, iface: T) -> Result<bool, ServiceError>
+    where
+        T: zbus::Interface,
+    {
+        let object_server = self.connection.object_server();
+        Ok(object_server.at(path.clone(), iface).await?)
+    }
+}
+
+/// Objects paths for known devices and connections
+#[derive(Debug, Default)]
+pub struct ObjectsRegistry {
+    pub devices: HashMap<String, String>,
+    pub connections: HashMap<Uuid, String>,
+}
+
+impl ObjectsRegistry {
+    /// Registers a network device.
+    ///
+    /// * `name`: device name.
+    /// * `path`: object path.
+    pub fn register_device(&mut self, name: &str, path: &str) {
+        self.devices.insert(name.to_string(), path.to_string());
+    }
+
+    /// Registers a network connection.
+    ///
+    /// * `uuid`: connection UUID.
+    /// * `path`: object path.
+    pub fn register_connection(&mut self, uuid: Uuid, path: &str) {
+        self.connections.insert(uuid, path.to_string());
+    }
+
+    /// Returns the path for a connection.
+    ///
+    /// * `uuid`: connection UUID.
+    pub fn connection_path(&self, uuid: Uuid) -> Option<&str> {
+        self.connections.get(&uuid).map(|p| p.as_str())
+    }
+
+    /// Deregisters a network connection.
+    ///
+    /// * `uuid`: connection UUID.
+    pub fn deregister_connection(&mut self, uuid: Uuid) -> Option<String> {
+        self.connections.remove(&uuid)
+    }
+
+    /// Returns all devices paths.
+    pub fn devices_paths(&self) -> Vec<String> {
```

Comment:
We should be using specialized strings that are guaranteed to be valid object paths.
The [zbus_names](https://docs.rs/zbus_names/2.5.0/zbus_names/index.html) crate exists exactly for this purpose and... it does not include object paths... because they appear on the wire so [the type lives in `zvariant`](https://docs.rs/zvariant/3.12.0/zvariant/struct.ObjectPath.html).

So,
```rs
use zvariant::ObjectPath;
    pub fn devices_paths(&self) -> Vec<OwnedObjectPath> {
or
    pub fn devices_paths(&self) -> Vec<ObjectPath> {
```

here and in many other places.
With respect to the D-Bus API it is an implementation detail, so we can leave it for later.

Reply:
I agree.